src/Compilers/Core/MSBuildTask/ManagedCompiler.cs

```diff
@@ -278,6 +283,169 @@ protected override Encoding StandardOutputEncoding
 
         #endregion
 
+        internal abstract BuildProtocolConstants.RequestLanguage Language { get; }
+
+        protected override int ExecuteTool(string pathToTool, string responseFileCommands, string commandLineCommands)
+        {
+            if (!UseSharedCompilation)
+            {
+                return base.ExecuteTool(pathToTool, responseFileCommands, commandLineCommands);
+            }
+
+            using (_sharedCompileCts = new CancellationTokenSource())
+            {
+                try
+                {
+                    var responseTask = BuildClient.TryRunServerCompilation(
+                        Language,
+                        CurrentDirectoryToUse(),
+                        GetArguments(commandLineCommands, responseFileCommands),
+                        _sharedCompileCts.Token,
+                        libEnvVariable: LibDirectoryToUse());
+
+                    responseTask.Wait(_sharedCompileCts.Token);
+
+                    ExitCode = responseTask.Result != null
+                        ? HandleResponse(responseTask.Result)
+                        : base.ExecuteTool(pathToTool, responseFileCommands, commandLineCommands);
+                }
+                catch (OperationCanceledException)
+                {
+                    ExitCode = 0;
+                }
+                catch (Exception e)
+                {
+                    Log.LogErrorWithCodeFromResources("Compiler.UnexpectedException");
+                    LogErrorOutput(e.ToString());
+                    ExitCode = -1;
+                }
+            }
+            return ExitCode;
+        }
+
+        /// <summary>
+        /// Cancel the in-process build task.
+        /// </summary>
+        public override void Cancel()
+        {
+            base.Cancel();
+
+            _sharedCompileCts?.Cancel();
+        }
+
+        /// <summary>
+        /// Get the current directory that the compiler should run in.
+        /// </summary>
+        private string CurrentDirectoryToUse()
+        {
+            // ToolTask has a method for this. But it may return null. Use the process directory
+            // if ToolTask didn't override. MSBuild uses the process directory.
+            string workingDirectory = GetWorkingDirectory();
+            if (string.IsNullOrEmpty(workingDirectory))
+                workingDirectory = Environment.CurrentDirectory;
```

Comment:
It's might be a matter of style, but I prefer the return a value explcitly and immediately once I know it is the intended return value, rather than updating a variable, and let control flow go to a 'return' statement below. So, I'd rewrite it as `return Environment.CurrentDirectory;`